[PT2][memory] correct wait tensor output size

xuanzhang816 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153569

This PR correctly handles the output buffer size of wait tensor nodes. 
![image](https://github.com/user-attachments/assets/fdcc5eb7-58cf-42a2-84b2-ce949cb9db92)

See [this doc](https://docs.google.com/document/d/1lkKulwIb-fYL_p8jn1SD6Lh1PoAKBgpBsU5sAH80leI/edit?tab=t.0#bookmark=id.w3n4k1y4rdz8) with testing details [internal only]

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

xuanzhang816 committed:
Update

[ghstack-poisoned]

xuanzhang816 commented:
@pytorchbot label "topic: not user facing"

eellison commented on torch/_inductor/memory.py:
```diff
@@ -193,6 +193,12 @@ def assign_memory_planning_info_for_scheduler_buffers(
         for dep in node.unmet_dependencies:
             dep_name_to_succ_nodes[dep.name].add(node)
+            # add successors due to mutations
+            if dep.name in name_to_buf:
+                buf = name_to_buf[dep.name]
+                if buf.get_mutations():
+                    mutated_buf_name = buf.get_mutations()[0]
+                    dep_name_to_succ_nodes[mutated_buf_name].add(node)
```
These should already be modeled via weak deps. can we use those instead ?

xuanzhang816 replied:
Do you have a code pointer for get to a node's weak dependency? Seems that by the time of the memory reorder pass, weak dependencies are pruned?

eellison replied:
Could we re-add them generically so that they work for all of the reordering passes, not just memory ? cc @BoyuanFeng who ran into this recently as well.

BoyuanFeng replied:
Could you add in [compute_dependencies()](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/scheduler.py#L2225)?
Also, what is the unit test to repro the issue?

xuanzhang816 replied:
Hmmm, I was not aware of this function, thanks for the link and let me check on this. 

For repro and unit test, I don't have one yet. I will write one that is related to mutations.

xuanzhang816 replied:
I updated the PR, the title and the description. 

My understanding of mutations is very wrong before so the issue is actually not about incorrect data dependency due to mutations, but that specifically for wait tensors, they are treated as mutations, but the layout makes it so that the buffer size is evaluated to be zero. The fix is that the size of wait tensor's output is inherited from the buffer that it mutates.

xuanzhang816 committed:
Update on "[PT2][memory] add missing dependencies due to mutations"


This PR adds the missing data dependency due to mutations (illustrated in the figure below).
![image](https://github.com/user-attachments/assets/38b629fa-56d3-40e3-879f-cd101adce034)


See [this doc](https://docs.google.com/document/d/1lkKulwIb-fYL_p8jn1SD6Lh1PoAKBgpBsU5sAH80leI/edit?tab=t.0#bookmark=id.w3n4k1y4rdz8) with testing details [internal only]

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

xuanzhang816 committed:
Update on "[PT2][memory] add missing dependencies due to mutations"


This PR adds the missing data dependency due to mutations (illustrated in the figure below).
![image](https://github.com/user-attachments/assets/38b629fa-56d3-40e3-879f-cd101adce034)


See [this doc](https://docs.google.com/document/d/1lkKulwIb-fYL_p8jn1SD6Lh1PoAKBgpBsU5sAH80leI/edit?tab=t.0#bookmark=id.w3n4k1y4rdz8) with testing details [internal only]

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

xuanzhang816 committed:
Update on "[PT2][memory] add missing dependencies due to mutations"


This PR adds the missing data dependency due to mutations (illustrated in the figure below).
![image](https://github.com/user-attachments/assets/38b629fa-56d3-40e3-879f-cd101adce034)


See [this doc](https://docs.google.com/document/d/1lkKulwIb-fYL_p8jn1SD6Lh1PoAKBgpBsU5sAH80leI/edit?tab=t.0#bookmark=id.w3n4k1y4rdz8) with testing details [internal only]

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
